File: src/main/java/com/intetics/lukyanenko/service/ShopService.java
package com.intetics.lukyanenko.service;

import com.intetics.lukyanenko.dao.*;
import com.intetics.lukyanenko.models.*;
import com.intetics.lukyanenko.web.AppService;
import javafx.util.Pair;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.DependsOn;
import org.springframework.stereotype.Service;

import javax.inject.Provider;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Map;

@Service
@DependsOn(value = "flyway")
public class ShopService implements AppService
{
  @Autowired
  Provider<Order>         orderProvider;
  @Autowired
  Provider<OrderDetail>   orderDetailProvider;
  @Autowired
  Provider<Customer>      customerProvider;
  @Autowired
  Provider<GoodsCategory> goodsCategoryProvider;
  @Autowired
  Provider<GoodsItem>     goodsItemProvider;
  @Autowired
  Provider<AppUser>       appUserProvider;
  
  private GoodsCategory newGoodsCategory()
  {
    return goodsCategoryProvider.get();
  }
  private GoodsItem     newGoodsItem()
  {
    return goodsItemProvider.get();
  }
  private Order         newOrder()
  {
    return orderProvider.get();
  }
  private OrderDetail   newOrderDetail()
  {
    return orderDetailProvider.get();
  }
  private AppUser       newAppUser()
  {
    return appUserProvider.get();
  }
  private Customer      newCustomer()
  {
    return customerProvider.get();
  }
  
  private final DAOFactory factory;
  
  @Autowired
  public ShopService(DAOFactory factory) {
    this.factory = factory;
  }
  
  @Override
  public List<AppUser> getAppUserList()
  {
    return factory.getDAO(AppUser.class).findAll();
  }
  
  @Override
  public AppUser getAppUserInfo(String name)
  {
    return ((AppUserDAO)factory.getDAO(AppUser.class)).getByName(name);
  }
  
  @Override
  public void setAppUser(AppUser appUser)
  {
    if (appUser.getIsNew())
      factory.getDAO(AppUser.class).add(appUser);
    else
      factory.getDAO(AppUser.class).update(appUser);
  }
  
  public void deleteAppUser(String appUserName)
  {
    factory.getDAO(AppUser.class).delete(new AppUser(appUserName));
  }
  
  @Override
  public List<GoodsCategory> getGoodCategories()
  {
    return factory.getDAO(GoodsCategory.class).findAll();
  }
  
  @Override
  public List<GoodsCategory> getGoodItemCategories(Integer goodsItemId)
  {
    return ((GoodsCategoryDAO)factory.getDAO(GoodsCategory.class)).getGoodsItemCategories(goodsItemId);
  }
  
  @Override
  public GoodsCategory getGoodCategory(Integer id)
  {
    return factory.getDAO(GoodsCategory.class).getByID(id);
  }
  
  @Override
  public void setGoodsCategory(GoodsCategory goodCategory)
  {
    if (goodCategory.getId() == null)
      factory.getDAO(GoodsCategory.class).add(goodCategory);
    else
      factory.getDAO(GoodsCategory.class).update(goodCategory);
  }
  
  @Override
  public void deleteGoodsCategory(Integer id)
  {
    GoodsCategory object = newGoodsCategory();
    object.setId(id);
    factory.getDAO(GoodsCategory.class).delete(object);
  }
  
  @Override
  public GoodsItem getGoodsItem(Integer id)
  {
    return factory.getDAO(GoodsItem.class).getByID(id);
  }
  
  @Override
  public GoodsItem getGoodsItemForEdit(Integer id)
  {
    GoodsItem object = getGoodsItem(id);
    object.setCategories(((GoodsCategoryDAO)factory.getDAO(GoodsCategory.class)).getGoodsItemCategories(id));
    return object;
  }
  
  @Override
  public GoodsItem getEmptyGoodsItem()
  {
    GoodsItem object = newGoodsItem();
    object.setCategories(new ArrayList<>(0));
    return object;
  }
  
  @Override
  public void setGoodsItem(GoodsItem goodsItem)
  {
    if (goodsItem.getId() != null)
      factory.getDAO(GoodsItem.class).update(goodsItem);
    else
      factory.getDAO(GoodsItem.class).add(goodsItem);
  }
  
  @Override
  public void deleteGoodsItem(Integer id)
  {
    GoodsItem object = newGoodsItem();
    object.setId(id);
    factory.getDAO(GoodsItem.class).delete(object);
  }
  
  @Override
  public List<GoodsItem> searchGoodItems(Double priceFrom, Double priceTill, String category, String itemName)
  {
    GoodsItemDAO dao = (GoodsItemDAO) factory.getDAO(GoodsItem.class);
    ArrayList<Pair<String, Pair<String, Object>>> params = new ArrayList<>();
    if (priceFrom != null)
      params.add(new Pair<>("price", new Pair<>(">= $(param)", priceFrom)));
    if (priceTill != null)
      params.add(new Pair<>("price", new Pair<>("<= $(param)", priceTill)));
    if (category != null && !category.isEmpty())
      params.add(new Pair<>("category", new Pair<>("like '%' || $(param) || '%'", category)));
    if (itemName != null && !itemName.isEmpty())
      params.add(new Pair<>("name", new Pair<>("like '%' || $(param) || '%'", itemName)));
    return dao.selectWhere(params);
  }
  
  @Override
  public Order getBasket(String sessionId)
  {
    OrderDAO dao = (OrderDAO)factory.getDAO(Order.class);
    Customer customer = getSessionCustomer(sessionId);
    Order basket = dao.findBasket(customer.getId());
    if (basket == null)
    {
      basket = newOrder();
      basket.setCreated(new Date());
      basket.setCustomer(customer);
      basket.setIsBasket(true);
      dao.add(basket);
    }
    return basket;
  }
  
  private Customer getSessionCustomer(String sessionId)
  {
    CustomerDAO dao = (CustomerDAO)factory.getDAO(Customer.class);
    Customer customer = dao.find(sessionId);
    if (customer == null)
    {
      customer = newCustomer();
      customer.setAnonymousSessionID(sessionId);
      dao.add(customer);
    }
    return customer;
  }
  
  @Override
  public void updateBasket(Order basket, ArrayList<OrderDetail> updates)
  {
    if (basket.getDetails() == null)
      fillOrderDetails(basket);
    //
    OrderDetailDAO dao = (OrderDetailDAO)factory.getDAO(OrderDetail.class);
    for(OrderDetail upd : updates)
      for(OrderDetail row : basket.getDetails())
        if (row.getId() == upd.getId())
          if (upd.getQuantity() == 0)
          {
            dao.delete(upd);
            basket.getDetails().remove(row);
          }
          else
          {
            row.setQuantity(upd.getQuantity());
            dao.update(row);
          }
  }
  
  @Override
  public void checkIfBasketCanBeCheckedOut(Order basket) throws CustomerIsAnonymous, IllegalArgumentException
  {
    if (!basket.getIsBasket())
      throw new IllegalArgumentException("This is not a basket");
    fillOrderDetails(basket);
    if (basket.getDetails().size() == 0)
      throw new IllegalArgumentException("Basket is empty");
    if (basket.getCustomer().getIsAnonymous())
      throw new CustomerIsAnonymous();
  }
  
  @Override
  public void addGoodsItemToBasket(Integer goodsItemid, String sessionId)
  {
    Order basket = getBasket(sessionId);
    OrderDetailDAO dao = (OrderDetailDAO)factory.getDAO(OrderDetail.class);
    OrderDetail detail = dao.findGoodsItemInOrder(basket, goodsItemid);
    if (detail == null)
    {
      detail = newOrderDetail();
      detail.setOrder(basket);
      detail.setGoodsItem(factory.getDAO(GoodsItem.class).getByID(goodsItemid));
      detail.setQuantity(1);
      dao.add(detail);
    }
    else
    {
      detail.setQuantity(detail.getQuantity() + 1);
      dao.update(detail);
    }
  }
  
  @Override
  public void deleteOrder(Integer id)
  {
    Order object = newOrder();
    object.setId(id);
    factory.getDAO(Order.class).delete(object);
  }
  
  @Override
  public Object getOrder(Integer id)
  {
    return factory.getDAO(Order.class).getByID(id);
  }
  
  @Override
  public void updateOrder(Integer id, Map<String, String> params)
  {
    
  }
  
  @Override
  public List<Order> getOrderList(Customer customer)
  {
    return null;
  }
  
  @Override
  public List<Customer> getCustomerList()
  {
    return null;
  }
  
  @Override
  public Customer getCustomer(Integer id)
  {
    return null;
  }
  
  @Override
  public void setCustomer(Customer customer)
  {
    
  }
  
  @Override
  public void deleteCustomer(Integer id)
  {
    
  }
  
  @Override
  public void clearBasket(String sessionId)
  {
    factory.getDAO(Order.class).delete(getBasket(sessionId));
  }
  
  @Override
  public OrderDetail getOrderDetailEmpty()
  {
    return newOrderDetail();
  }
  
  @Override
  public Order makeOrderFromBasket(Order basket) throws CustomerIsAnonymous
  {
    checkIfBasketCanBeCheckedOut(basket);
    OrderDAO dao = (OrderDAO)factory.getDAO(Order.class);
    Order order = dao.convertBasketToOrder(basket);
    return order;
  }
  
  @Override
  public void fillOrderDetails(Order order)
  {
    if (order.getDetails() == null)
      order.setDetails(((OrderDetailDAO)factory.getDAO(OrderDetail.class))
                         .getListForOrder(order,
                                          new OrderDetailDAO.SubdetailsFillable() {
                                            @Override
                                            public void fill(OrderDetail detail, ResultSet resultSet,
                                                             String fieldNamePrefix
                                                            ) throws SQLException
                                            {
                                              detail.setGoodsItem(factory.getDAO(GoodsItem.class)
                                                                         .populateFromResultSet(resultSet, fieldNamePrefix)
                                                                 );
                                            }
                                          }));
  }
  
  @Override
  public AppUser getAppUserEmptyNew()
  {
    AppUser object = newAppUser();
    object.setIsNew(true);
    return object;
  }
  
  @Override
  public GoodsCategory getGoodCategoryEmpty()
  {
    return newGoodsCategory();
  }
  
}

File: src/main/java/com/intetics/lukyanenko/service/ShopService.kt
package com.intetics.lukyanenko.service

import com.intetics.lukyanenko.dao.*
import com.intetics.lukyanenko.models.*
import com.intetics.lukyanenko.web.AppService
import javafx.util.Pair
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.annotation.DependsOn
import org.springframework.stereotype.Service
import java.util.*
import javax.inject.Provider
import kotlin.collections.ArrayList

@Service
@DependsOn(value = "flyway")
class ShopService
@Autowired
constructor(private val factory: DAOFactory) : AppService {

    @Autowired
    internal lateinit var orderProvider: Provider<Order>
    @Autowired
    internal lateinit var orderDetailProvider: Provider<OrderDetail>
    @Autowired
    internal lateinit var customerProvider: Provider<Customer>
    @Autowired
    internal lateinit var goodsCategoryProvider: Provider<GoodsCategory>
    @Autowired
    internal lateinit var goodsItemProvider: Provider<GoodsItem>
    @Autowired
    internal lateinit var appUserProvider: Provider<AppUser>

    private fun newGoodsCategory(): GoodsCategory = goodsCategoryProvider.get()

    private fun newGoodsItem(): GoodsItem = goodsItemProvider.get()

    private fun newOrder(): Order = orderProvider.get()

    private fun newOrderDetail(): OrderDetail = orderDetailProvider.get()

    private fun newAppUser(): AppUser = appUserProvider.get()

    private fun newCustomer(): Customer = customerProvider.get()

    override fun getAppUserList(): List<AppUser> = factory.getDAO(AppUser::class).findAll()

    override fun getAppUserInfo(name: String): AppUser? = (factory.getDAO(AppUser::class) as AppUserDAO).getByName(name)

    override fun setAppUser(appUser: AppUser) {
        with(factory.getDAO(AppUser::class)) {
            if (appUser.isNew)
                add(appUser)
            else
                update(appUser)
        }
    }

    override fun deleteAppUser(appUserName: String) =
            factory.getDAO(AppUser::class).delete(AppUser(appUserName))

    override fun getGoodCategories(): List<GoodsCategory> =
            factory.getDAO(GoodsCategory::class).findAll()

    override fun getGoodItemCategories(goodsItemId: Int): List<GoodsCategory> =
            (factory.getDAO(GoodsCategory::class) as GoodsCategoryDAO).getGoodsItemCategories(goodsItemId)

    override fun getGoodCategory(id: Int): GoodsCategory? = factory.getDAO(GoodsCategory::class).getByID(id)

    override fun setGoodsCategory(goodCategory: GoodsCategory) {
        with(factory.getDAO(GoodsCategory::class)) {
            if (goodCategory.id == null)
                add(goodCategory)
            else
                update(goodCategory)
        }
    }

    override fun deleteGoodsCategory(id: Int) {
        factory.getDAO(GoodsCategory::class).delete(newGoodsCategory().also { it.id = id })
    }

    override fun getGoodsItem(id: Int): GoodsItem? {
        return factory.getDAO(GoodsItem::class).getByID(id)
    }

    override fun getGoodsItemForEdit(id: Int): GoodsItem? {
        return getGoodsItem(id)?.also {
            it.categories = (factory.getDAO(GoodsCategory::class) as GoodsCategoryDAO).getGoodsItemCategories(id)
        }
    }

    override fun getEmptyGoodsItem(): GoodsItem = newGoodsItem()

    override fun setGoodsItem(goodsItem: GoodsItem) {
        with(factory.getDAO(GoodsItem::class)) {
            if (goodsItem.id != null)
                update(goodsItem)
            else
                add(goodsItem)
        }
    }

    override fun deleteGoodsItem(id: Int) {
        factory.getDAO(GoodsItem::class).delete(newGoodsItem().also { it.id = id })
    }

    override fun searchGoodItems(priceFrom: Double?, priceTill: Double?, category: String?, itemName: String?): List<GoodsItem> {
        val params = ArrayList<Pair<String, Pair<String, Any>>>()
        if (priceFrom != null)
            params.add(Pair("price", Pair<String, Any>(">= $(param)", priceFrom)))
        if (priceTill != null)
            params.add(Pair("price", Pair<String, Any>("<= $(param)", priceTill)))
        if (!category.isNullOrEmpty())
            params.add(Pair("category", Pair<String, Any>("like '%' || $(param) || '%'", category)))
        if (!itemName.isNullOrEmpty())
            params.add(Pair("name", Pair<String, Any>("like '%' || $(param) || '%'", itemName)))
        return (factory.getDAO(GoodsItem::class) as GoodsItemDAO).selectWhere(params)
    }

    override fun getBasket(sessionId: String, withDetails: Boolean): Order {
        val dao = factory.getDAO(Order::class) as OrderDAO
        val customer = getSessionCustomer(sessionId)
        return dao.findBasket(customer.id!!)?.also { if (withDetails) fillOrderDetails(it) } ?:
                newOrder()
                        .apply {
                            this.created = Date()
                            this.customer = customer
                            this.isBasket = true
                            this.details = mutableListOf()
                            dao.add(this)
                        }
    }

    private fun getSessionCustomer(sessionId: String): Customer {
        val dao = factory.getDAO(Customer::class) as CustomerDAO
        return dao.find(sessionId) ?: newCustomer()
                .also {
                    it.anonymousSessionID = sessionId
                    dao.add(it)
                }
    }

    override fun updateBasket(basket: Order, updates: ArrayList<OrderDetail>) {
        basket.details = getOrderDetails(basket).apply {
            val dao = factory.getDAO(OrderDetail::class) as OrderDetailDAO
            val userWishes = updates.associateBy { it.id }
            forEach { basketRow ->
                userWishes[basketRow.id]?.let { wish ->
                    if (wish.quantity == 0.0) {
                        dao.delete(wish)
                        remove(basketRow)
                    } else {
                        basketRow.quantity = wish.quantity
                        dao.update(basketRow)
                    }
                }
            }
        }
    }

    override fun checkIfBasketCanBeCheckedOut(basket: Order) {
        if (!basket.isBasket)
            throw IllegalArgumentException("This is not a basket")
        fillOrderDetails(basket)
        if (basket.details?.isEmpty() != false)
            throw IllegalArgumentException("Basket is empty")
        if (basket.customer?.isAnonymous() != false)
            throw AppService.CustomerIsAnonymous()
    }

    override fun addGoodsItemToBasket(goodsItemId: Int, sessionId: String) {
        val basket = getBasket(sessionId)
        val dao = factory.getDAO(OrderDetail::class) as OrderDetailDAO
        dao.findGoodsItemInOrder(basket, goodsItemId)?.also {
            it.quantity++
            dao.update(it)
        } ?: newOrderDetail().also {
            it.order = basket
            it.goodsItem = factory.getDAO(GoodsItem::class).getByID(goodsItemId)
            it.quantity++
            dao.add(it)
        }
    }

    override fun deleteOrder(id: Int) {
        factory.getDAO(Order::class).delete(newOrder().also { it.id = id })
    }

    override fun getOrder(id: Int): Order? = factory.getDAO(Order::class).getByID(id)

    override fun updateOrder(id: Int, params: Map<String, String>) {
        // todo
    }

    override fun getOrderList(customer: Customer): List<Order> {
        // todo
        return ArrayList<Order>(0)
    }

    override fun getCustomerList(): List<Customer> {
        // todo
        return ArrayList<Customer>(0)
    }

    override fun getCustomer(id: Int): Customer? = factory.getDAO(Customer::class).getByID(id)

    override fun setCustomer(customer: Customer) {
        with(factory.getDAO(Customer::class)) {
            if (customer.id == null)
                add(customer)
            else
                update(customer)
        }
    }

    override fun deleteCustomer(id: Int) {
        factory.getDAO(Customer::class).delete(Customer(id))
    }

    override fun clearBasket(sessionId: String) {
        factory.getDAO(Order::class).delete(getBasket(sessionId))
    }

    override fun getOrderDetailEmpty(): OrderDetail = newOrderDetail()

    override fun makeOrderFromBasket(basket: Order): Order {
        checkIfBasketCanBeCheckedOut(basket)
        return (factory.getDAO(Order::class) as OrderDAO).convertBasketToOrder(basket)
    }

    override fun fillOrderDetails(order: Order) {
        order.details = getOrderDetails(order)
    }

    override fun getOrderDetails(order: Order): MutableList<OrderDetail> =
            order.details ?: (factory.getDAO(OrderDetail::class) as OrderDetailDAO).getListForOrder(order)

    override fun getAppUserEmptyNew(): AppUser = newAppUser().also { it.isNew = true }

    override fun getGoodCategoryEmpty(): GoodsCategory = newGoodsCategory()
}